crypto/modes/asm/aes-gcm-avx512.pl

```diff
@@ -0,0 +1,5141 @@
+  } else {
+
+    # ;; Encrypt case
+    # ;; - cipher blocks are in CTR0-CTR3
+    &ZMM_OPCODE3_DSTR_SRC1R_SRC2R_BLOCKS_0_16(
+      $NUM_BLOCKS, "vpshufb", $DAT0, $DAT1,     $DAT2,     $DAT3,     $CTR0,
+      $CTR1,       $CTR2,     $CTR3, $SHUFMASK, $SHUFMASK, $SHUFMASK, $SHUFMASK);
+  }
+
+  if ($INSTANCE_TYPE eq "multi_call") {
+
+    # ;; Extract the last block for partials and multi_call cases
+    if ($NUM_BLOCKS <= 4) {
+      $code .= "vextracti32x4     \$`($NUM_BLOCKS-1)`,$DAT0,$LAST_GHASH_BLK\n";
+    } elsif ($NUM_BLOCKS <= 8) {
+      $code .= "vextracti32x4     \$`($NUM_BLOCKS-5)`,$DAT1,$LAST_GHASH_BLK\n";
+    } elsif ($NUM_BLOCKS <= 12) {
+      $code .= "vextracti32x4     \$`($NUM_BLOCKS-9)`,$DAT2,$LAST_GHASH_BLK\n";
+    } else {
+      $code .= "vextracti32x4     \$`($NUM_BLOCKS-13)`,$DAT3,$LAST_GHASH_BLK\n";
+    }
+  }
+
+}
+
+# ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+# ;; Computes GHASH on 1 to 16 blocks
+sub INITIAL_BLOCKS_PARTIAL_GHASH {
+  my $AES_KEYS        = $_[0];     # [in] key pointer
+  my $GCM128_CTX      = $_[1];     # [in] context pointer
+  my $LENGTH          = $_[2];     # [in/clobbered] length in bytes
+  my $NUM_BLOCKS      = $_[3];     # [in] can only be 1, 2, 3, 4, 5, ..., 15 or 16 (not 0)
+  my $HASH_IN_OUT     = $_[4];     # [in/out] XMM ghash in/out value
+  my $ENC_DEC         = $_[5];     # [in] cipher direction (ENC/DEC)
+  my $INSTANCE_TYPE   = $_[6];     # [in] multi_call or single_call
+  my $DAT0            = $_[7];     # [in] ZMM with cipher text shuffled for GHASH
+  my $DAT1            = $_[8];     # [in] ZMM with cipher text shuffled for GHASH
+  my $DAT2            = $_[9];     # [in] ZMM with cipher text shuffled for GHASH
+  my $DAT3            = $_[10];    # [in] ZMM with cipher text shuffled for GHASH
+  my $LAST_CIPHER_BLK = $_[11];    # [in] XMM with ciphered counter block partially xor'ed with text
+  my $LAST_GHASH_BLK  = $_[12];    # [in] XMM with last cipher text block shuffled for GHASH
+  my $ZT0             = $_[13];    # [clobbered] ZMM temporary
+  my $ZT1             = $_[14];    # [clobbered] ZMM temporary
+  my $ZT2             = $_[15];    # [clobbered] ZMM temporary
+  my $ZT3             = $_[16];    # [clobbered] ZMM temporary
+  my $ZT4             = $_[17];    # [clobbered] ZMM temporary
+  my $ZT5             = $_[18];    # [clobbered] ZMM temporary
+  my $ZT6             = $_[19];    # [clobbered] ZMM temporary
+  my $ZT7             = $_[20];    # [clobbered] ZMM temporary
+  my $ZT8             = $_[21];    # [clobbered] ZMM temporary
+  my $PBLOCK_LEN      = $_[22];    # [in] partial block length
+  my $GH              = $_[23];    # [in] ZMM with hi product part
+  my $GM              = $_[24];    # [in] ZMM with mid prodcut part
+  my $GL              = $_[25];    # [in] ZMM with lo product part
+
+  my $rndsuffix = &random_string();
+
+  if ($INSTANCE_TYPE eq "single_call") {
```

Comment:
All this block can be removed, as "single_call" is never set

Reply:
Please see answer about multi_call / single_call above.

Reply:
Removed.